crypto/threads_pthread.c

```diff
@@ -42,12 +45,562 @@
+        if (data->thread_qps[i].lock == lock) {
+            /*
+             * As with read side acquisition, we use __ATOMIC_RELEASE here
+             * to ensure that the decrement is published immediately 
+             * to any write side waiters
+             */
+            ATOMIC_SUB_FETCH(&data->thread_qps[i].qp->users, VAL_READER,
+                             __ATOMIC_RELEASE);
+            data->thread_qps[i].qp = NULL;
+            data->thread_qps[i].lock = NULL;
+            return;
+        }
+    }
+}
+
+/*
+ * Write side allocation routine to get the current qp
+ * and replace it with a new one
+ */
+static struct rcu_qp *update_qp(CRYPTO_RCU_LOCK lock)
+{
+    uint64_t new_id;
+    uint64_t current_idx;
+
+    pthread_mutex_lock(&lock->alloc_lock);
+
+    /*
+     * we need at least one qp to be available with one
+     * left over, so that readers can start working on
+     * one that isn't yet being waited on
+     */
+    while (lock->group_count - lock->writers_alloced < 2)
+        /* we have to wait for one to be free */
+        pthread_cond_wait(&lock->alloc_signal, &lock->alloc_lock);
+
+    current_idx = lock->current_alloc_idx;
+
+    /* Allocate the qp */
+    lock->writers_alloced++;
+
+    /* increment the allocation index */
+    lock->current_alloc_idx =
+        (lock->current_alloc_idx + 1) % lock->group_count;
+
+    /* get and insert a new id */
+    new_id = lock->id_ctr;
+    lock->id_ctr++;
+
+    new_id = VAL_ID(new_id);
+    /*
+     * Even though we are under a write side lock here
+     * We need to use atomic instructions to ensure that the results
+     * of this update are published to the read side prior to updating the
+     * reader idx below
+     */
+    ATOMIC_AND_FETCH(&lock->qp_group[current_idx].users, ID_MASK,
+                     __ATOMIC_RELEASE);
+    ATOMIC_OR_FETCH(&lock->qp_group[current_idx].users, new_id,
+                    __ATOMIC_RELEASE);
+
+    /*
+     * update the reader index to be the prior qp
+     * Note the use of __ATOMIC_RELEASE here is based on the corresponding use
+     * of __ATOMIC_ACQUIRE in get_hold_current_qp, as we wan't any publication
+     * of this value to be seen on the read side immediately after it happens
+     */
+    ATOMIC_STORE_N(&lock->reader_idx, lock->current_alloc_idx,
+                   __ATOMIC_RELEASE);
+
+    /* wake up any waiters */
+    pthread_cond_signal(&lock->alloc_signal);
+    pthread_mutex_unlock(&lock->alloc_lock);
+    return &lock->qp_group[current_idx];
+}
+
+static void retire_qp(CRYPTO_RCU_LOCK lock, struct rcu_qp *qp)
+{
+    pthread_mutex_lock(&lock->alloc_lock);
+    lock->writers_alloced--;
+    pthread_cond_signal(&lock->alloc_signal);
+    pthread_mutex_unlock(&lock->alloc_lock);
+}
+
+static struct rcu_qp *allocate_new_qp_group(CRYPTO_RCU_LOCK lock,
+                                            int count)
+{
+    struct rcu_qp *new =
+        OPENSSL_zalloc(sizeof(*new) * count);
```

Comment:
What if this allocation fails? Is it still correct to set `lock->group_count` below?